### PR TITLE
Adding kustomization files for remote bases

### LINKED ISSUE #2391: Feature Request: Add support for kustomize remote-base
**What did you do?**
N/A

**What did you expect to see?**
N/A

**What did you see instead? Under which circumstances?**
N/A

**Environment**
N/A

I just wanted to submit a feature request to add a kustomization.yaml file to your kube-prometheus [manifests](https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus/manifests) directory.  This would allow Kustomize users to use a [remote-base](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/remoteBuild.md) and just maintain their overlays locally.  Then updating would be as simple as changing the ref version in the overlay file instead of downloading, extraction and moving files into a local base directory.

If you understandably don't want the kustomization.yaml file in the manifest dir, you could also just have a kustomize directory in [kube-prometheus](https://github.com/coreos/prometheus-operator/tree/master/contrib/kube-prometheus) that has nothing but the kustomization.yaml file, which points to all the manifest files.

### PR DESCRIPTION
**Pull request by coryodaniel**
Closes #2391

Adds support to allow kustomize users to point to a [remote
base](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/remoteBuild.md) of the prometheus-operator and kube-prometheus. Now that kustomize is a part of kubectl, this is provides a very simple way of installing prometheus-operator and kube-prometheus with no additional tools besides kubectl.

* [Added] kustomization.yaml: hardcoded kustomization base for prometheus-operator
bundle.yaml
* [Added] kustomization generation from kube-prometheus example jsonnet

### CODE REVIEW

**Comment by coryodaniel** [comment]
First time jsonnetting. I feel like I did this wrong, but the result is as intended :D

Direction welcome.

File: kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Comment (brancz) [inline kustomization.yaml]: How did this file get here? It doesn't look generated.

File: contrib/kube-prometheus/example.jsonnet [hunk 2]
```diff
@@ -5,11 +5,28 @@ local kp =
...
 
-{ ['00namespace-' + name]: kp.kubePrometheus[name] for name in std.objectFields(kp.kubePrometheus) } +
-{ ['0prometheus-operator-' + name]: kp.prometheusOperator[name] for name in std.objectFields(kp.prometheusOperator) } +
-{ ['node-exporter-' + name]: kp.nodeExporter[name] for name in std.objectFields(kp.nodeExporter) } +
-{ ['kube-state-metrics-' + name]: kp.kubeStateMetrics[name] for name in std.objectFields(kp.kubeStateMetrics) } +
-{ ['alertmanager-' + name]: kp.alertmanager[name] for name in std.objectFields(kp.alertmanager) } +
-{ ['prometheus-' + name]: kp.prometheus[name] for name in std.objectFields(kp.prometheus) } +
-{ ['prometheus-adapter-' + name]: kp.prometheusAdapter[name] for name in std.objectFields(kp.prometheusAdapter) } +
-{ ['grafana-' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) }
+local kustomization = {
```
Comment (brancz) [inline contrib/kube-prometheus/example.jsonnet]: To deduplicate the naming scheme here, let's move the `manifests` variable above the `kustomization` one, then this object definition suddenly becomes very small (I haven't tried this but it should be something close to this):

```jsonnet
local manifests =
  { ['00namespace-' + name]: kp.kubePrometheus[name] for name in std.objectFields(kp.kubePrometheus) } +
  { ['0prometheus-operator-' + name]: kp.prometheusOperator[name] for name in std.objectFields(kp.prometheusOperator) } +
  { ['node-exporter-' + name]: kp.nodeExporter[name] for name in std.objectFields(kp.nodeExporter) } +
  { ['kube-state-metrics-' + name]: kp.kubeStateMetrics[name] for name in std.objectFields(kp.kubeStateMetrics) } +
  { ['alertmanager-' + name]: kp.alertmanager[name] for name in std.objectFields(kp.alertmanager) } +
  { ['prometheus-' + name]: kp.prometheus[name] for name in std.objectFields(kp.prometheus) } +
  { ['prometheus-adapter-' + name]: kp.prometheusAdapter[name] for name in std.objectFields(kp.prometheusAdapter) } +
  { ['grafana-' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) };

local kustomization = {
  apiVersion: 'kustomize.config.k8s.io/v1beta1',
  kind: 'Kustomization',
  resources: std.map(function(name) { name + '.yaml' } , std.objectFields(manifests),
};

manifests { kustomization: kustomization }
```

File: kustomization.yaml (hunk 1 above)
Comment (coryodaniel) [reply kustomization.yaml]: That one is hard coded! I dropped it there for anyone wanting to use the bundle.yaml as their kustomize base.

**Comment by coryodaniel** [comment]
Refactored the resource/manifests names.

@brancz re:
>  Could you also instead of modifying the example.jsonnet just move this to contrib/kube-prometheus/examples/kustomization.jsonnet and change the script called for rendering to that one instead of the example.jsonnet

Yep, do you want me to revert to the original `example.jsonnet` file as well?

**Comment by coryodaniel** [comment]
* [Modified] Makefile to run kustomize.jsonnet
* [Modified] Moved kustomization from example.jsonnet to
examples/kustomize.jsonnet
* [Modified] kustomization file to land at root of kube-prometheus so
current users can continue to use `kubectl apply -f manifests`

File: contrib/kube-prometheus/example.jsonnet [hunk 3]
```diff
@@ -12,4 +12,4 @@ local kp =
 { ['alertmanager-' + name]: kp.alertmanager[name] for name in std.objectFields(kp.alertmanager) } +
 { ['prometheus-' + name]: kp.prometheus[name] for name in std.objectFields(kp.prometheus) } +
 { ['prometheus-adapter-' + name]: kp.prometheusAdapter[name] for name in std.objectFields(kp.prometheusAdapter) } +
-{ ['grafana-' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) }
+{ ['grafana-' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) }
```
Comment (metalmatze) [inline contrib/kube-prometheus/example.jsonnet]: Add a newline at the end. Thanks.

File: contrib/kube-prometheus/examples/kustomize.jsonnet [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+local kustomizationResourceFile(name) = "./manifests/" + name + ".yaml";
+local kustomization = {
+  apiVersion: 'kustomize.config.k8s.io/v1beta1',
+  kind: 'Kustomization',
+  resources: std.map(kustomizationResourceFile, std.objectFields(manifests))
+};
+
+manifests {
+  "../kustomization": kustomization,
+}
```
Comment (metalmatze) [inline contrib/kube-prometheus/examples/kustomize.jsonnet]: Add a newline at the end. Thanks.

File: kustomization.yaml [hunk 5]
```diff
@@ -0,0 +1,4 @@
...
+kind: Kustomization
+resources:
+- bundle.yaml
```
Comment (metalmatze) [inline kustomization.yaml]: Add a newline at the end. Thanks.

**Comment by brancz** [comment]
this needs a docs regen, then should be good I think

**Comment by coryodaniel** [comment]
Whats the command for that? When I `make **.md` from contrib/kube-prometheus the only changes I see are newlines on:

contrib/kube-prometheus/README.md
contrib/kube-prometheus/docs/developing-prometheus-rules-and-grafana-dashboards.md

**Comment by coryodaniel** [comment]
@brancz feeling like a huge newb over here 😅😅

**Comment by brancz** [comment]
lgtm :+1: thanks for bearing with us! :)

**Comment by coryodaniel** [comment]
No problem at all. I appreciate the guidance!

**Pull request merged**